<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="UTF-8">
    <title>PARE by spundhir</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="stylesheets/normalize.css" media="screen">
    <link href='https://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">
  </head>
  <body>
    <section class="page-header">
      <h1 class="project-name">PARE</h1>
      <h2 class="project-tagline">PARE: a computational pipeline to Predict Active Regulatory Elements</h2>
      <a href="https://github.com/spundhir/PARE" class="btn">View on GitHub</a>
      <a href="https://github.com/spundhir/PARE/zipball/master" class="btn">Download .zip</a>
      <a href="https://github.com/spundhir/PARE/tarball/master" class="btn">Download .tar.gz</a>
    </section>

    <section class="main-content">
      <p>PARE is a computational method to predict active enhancers and promoters. It exploits the peak-valley-peak (PVP) pattern by which histone modifications are enriched at these regulatory elements. Specifically, at enhancers and promoters, H3K4me1 and H3K4me3-based PVP patterns are observed, respectively.</p>

<h1>
<a id="description" class="anchor" href="#description" aria-hidden="true"><span class="octicon octicon-link"></span></a>Description</h1>

<pre><code>PARE is a computational pipeline to Predict Active Regulatory Elements (enhancers and promoters). It implements a novel approach to detect Peak-Valley-Peak (PVP) pattern defined based on H3K4me1 and H3K4me3 signal at enhancers and promoters, respectively.

- Genomic regions enriched for H3K4me1 PVP pattern are predicted as active enhancers.

- Genomic regions enriched for H3K4me3 PVP pattern are predicted as active promoters.
</code></pre>

<h1>
<a id="version" class="anchor" href="#version" aria-hidden="true"><span class="octicon octicon-link"></span></a>Version</h1>

<pre><code>0.02
</code></pre>

<h1>
<a id="programs-and-datasets" class="anchor" href="#programs-and-datasets" aria-hidden="true"><span class="octicon octicon-link"></span></a>Programs and datasets</h1>

<pre><code>Generally, the user would be interested in following two scripts:

- pare: it is the main script to detect enhancers or promoters based on input BAM files (H3K4me1 - enhancers; H3K4me3 - promoters).

- bed2direction: this script is used to detect directionality of stable transcription at promoter regions, provided as input in BED format.

An example dataset and expected results are available at:

- http://servers.binf.ku.dk/pare/download/test_run/
</code></pre>

<h1>
<a id="installation" class="anchor" href="#installation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Installation</h1>

<pre><code>1. To install PARESuite, download PARESuite.tar.gz and unpack it. A directory, PARESuite will be created

    tar -zxvf PARESuite.tar.gz

2. Now compile and create executable blockbuster

    make or make all

3. Export environment variable 'PAREPATH' containing path to PARESuite installation directory

    export PAREPATH=&lt;path to PARESuite installation directory&gt;

4. Add 'PAREPATH' to your 'PATH' environment variable

    export PATH=$PATH:$PAREPATH/bin

5. Add 'PAREPATH' to your 'PERL5LIB' environment variable

    export PERL5LIB=$PERL5LIB:$PAREPATH/share/perl/

To permanently add or update the environment variable(s), add the last three export commands in your ~/.bashrc file
</code></pre>

<h1>
<a id="dependency" class="anchor" href="#dependency" aria-hidden="true"><span class="octicon octicon-link"></span></a>Dependency</h1>

<pre><code>We assume that the following programming platforms are installed and working: perl, R, and gcc. Besides, following packages should be installed.

1. Install the needed perl modules

    sudo cpan Tie::IxHash Statistics::Basic

2. R modules are installed by entering R (type R on the cmdline) and then enter the following three commands (follow the instructions on the screen):

    install.packages(c("ggplot2", "gridExtra", "optparse", "randomForest", "e1071"))

    source("http://bioconductor.org/biocLite.R")

    biocLite(c("DESeq"))

3. download samtools from http://sourceforge.net/projects/samtools/files/samtools/1.2/samtools-1.2.tar.bz2/download, go to the download location and do

    tar xjf samtools-1.2.tar.bz2

    cd samtools-1.2

    make -j10 prefix=$HOME install

4. download bedtools from https://github.com/arq5x/bedtools2/releases/download/v2.23.0/bedtools-2.23.0.tar.gz, go to the download location and do

    tar xzf BEDTools.v2.23

    cd bedtools-2.23.0/

    make -j 10

    cp bin/* $HOME/bin

5. download featureCounts (subread) from http://sourceforge.net/projects/subread/files/subread-1.4.6-p4/, go to the download location and do

    tar xzf subread-1.4.6-p4-Linux-x86_64.tar.gz

    cd subread-1.4.6-p3-Linux-x86_64

    cp bin/featureCounts $HOME/bin

6. download bedGraphToBigWig from http://hgdownload.soe.ucsc.edu/admin/exe/ for your operating system, go to the download location and do

    cp bedGraphToBigWig $HOME/bin

    chmod 755 $HOME/bin/bedGraphToBigWig
</code></pre>

<h1>
<a id="usage" class="anchor" href="#usage" aria-hidden="true"><span class="octicon octicon-link"></span></a>Usage</h1>

<pre><code>PARESuite is called with the following parameters

pare -i &lt;BAM file (rep1)&gt; -j &lt;BAM file (rep2)&gt; [OPTIONS]
</code></pre>

<h1>
<a id="example" class="anchor" href="#example" aria-hidden="true"><span class="octicon octicon-link"></span></a>Example</h1>

<pre><code>An usage example of PARESuite is shown below. As input, the pipeline requires mapped reads in BAM format. An example dataset and expected results are available at http://servers.binf.ku.dk/pare/download/test_run/

pare -i data/h3k4me1_helas3_Rep1.bam -j data/h3k4me1_helas3_Rep2.bam -k data/optimal.h3k4me1_helas3_Rep0_Vs_control_helas3_Rep0.regionPeak.gz -o results -m hg19 -p &amp;&gt; pare.log
</code></pre>

<h1>
<a id="input" class="anchor" href="#input" aria-hidden="true"><span class="octicon octicon-link"></span></a>Input</h1>

<pre><code>As input, the pipeline requires mapped reads in BAM format. The name of the input files should be formatted as

Input file name (replicate 1): &lt;unique id&gt;&lt;Rep1&gt;.bam (example: h3k4me1_Rep1.bam)

Input file name (replicate 2): &lt;unique id&gt;&lt;Rep2&gt;.bam (example: h3k4me1_Rep2.bam)

The chromosome identifier in the input BAM files should start with chr, for example as chrY and not like Y.
</code></pre>

<h1>
<a id="output" class="anchor" href="#output" aria-hidden="true"><span class="octicon octicon-link"></span></a>Output</h1>

<pre><code>The results from the PARESuite are presented in two text files:

a) RESULTS.TXT: main result file in BED format 

For easy access, the html version of this file (RESULTS.HTML) is also available within the output directory

b) RESULTS.UCSC: file to view the enhancer and promoter regions in UCSC browser
</code></pre>

<h1>
<a id="more-info" class="anchor" href="#more-info" aria-hidden="true"><span class="octicon octicon-link"></span></a>More info</h1>

<pre><code>for more and latest information, please refer to https://github.com/spundhir/pare or http://servers.binf.ku.dk/pare/ 
</code></pre>

<h1>
<a id="license" class="anchor" href="#license" aria-hidden="true"><span class="octicon octicon-link"></span></a>License</h1>

<pre><code>PARE: a computational pipeline to Predict Active Regulatory Elements using histone marks
Copyright (C) 2015  Sachin Pundhir (pundhir@binf.ku.dk)

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see &lt;http://www.gnu.org/licenses/&gt;.
</code></pre>

      <footer class="site-footer">
        <span class="site-footer-owner"><a href="https://github.com/spundhir/PARE">PARE</a> is maintained by <a href="https://github.com/spundhir">spundhir</a>.</span>

        <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the <a href="https://github.com/jasonlong/cayman-theme">Cayman theme</a> by <a href="https://twitter.com/jasonlong">Jason Long</a>.</span>
      </footer>

    </section>

  
  </body>
</html>
