<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="UTF-8">
    <title>PARE by spundhir</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="stylesheets/normalize.css" media="screen">
    <link href='https://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">
  </head>
  <body>
    <section class="page-header">
      <h1 class="project-name">PARE</h1>
      <h2 class="project-tagline">PARE: a computational method to <u>P</u>redict <u>A</u>ctive <u>R</u>egulatory <u>E</u>lements</h2>
      <a href="https://github.com/spundhir/PARE" class="btn">View on GitHub</a>
      <a href="https://github.com/spundhir/PARE/zipball/master" class="btn">Download .zip</a>
      <a href="https://github.com/spundhir/PARE/tarball/master" class="btn">Download .tar.gz</a>
    </section>

    <section class="main-content">
      <h1>
<a id="what" class="anchor" href="#what" aria-hidden="true"><span class="octicon octicon-link"></span></a>What is PARE?</h1>

<p>PARE is a computational method to <u><b>P</b></u>redict <u><b>A</b></u>ctive <u><b>R</b></u>egulatory <u><b>E</b></u>lements, specifically enhancers and promoters.

<h1>
<a id="why" class="anchor" href="#why" aria-hidden="true"><span class="octicon octicon-link"></span></a>Why use PARE?</h1>

<p>For predicting enhancers and promoters, PARE offers following advantages over available methods:</p>
<ol>
    <li>PARE uses only <u>single histone mark</u> (H3K4me1 or H3K4me3) for its predictions. Therefore, it can be used for cell types where limited ChIP-seq data is available. This may be due to low amount of sample or high cost.</li>
    <li>It predicts enhancers and promoters with <u>high specificity</u> for their activity. This is because PARE uses robust indicator of activity in the form of PVP pattern for its predictions.</li>
    <li>Due to above two advantages, PARE is very efficient to <u>study dynamics in the activity</u> of enhancers or promoters across different cell types or time points for which only H3K4me1/me3 data is available.</li>
    <li>PARE also includes a model to predict <u>directionality of stable transcription</u> at promoters. This is done using the same H3K4me1 and H3K4me3 data that was used for the prediction of enhancers and promoters, respectively.</li>
</ol>

<h1>
<a id="how" class="anchor" href="#how" aria-hidden="true"><span class="octicon octicon-link"></span></a>How it works?</h1>
PARE exploits two important properties of active enhancer and promoter regions, which are:</p>

<ol>
    <li>They are nucleosome depleted, therefore called as Nucleosome Free Regions (NFRs); and,</li>
    <li>The nucleosome flanking these regions are enriched for distinct histone modifications, which in case of enhancers is H3K4me1 and in case of promoters is H3K4me3.</li>
</ol>

<p>As a consequence of the above two properties, H3K4me1 and H3K4me3 signal is distinctly arranged in the form of a peak-valley-peak (PVP) pattern at enhancers and promoters, respectively. PARE, in essence, captures this PVP-pattern, and predicts:</p>

<ol>
    <li>Genomic regions enriched for H3K4me1 PVP pattern as active enhancers.</li>
    <li>Genomic regions enriched for H3K4me3 PVP pattern as active promoters.</li>
</ol>

<h1>
<a id="flowdiagram" class="anchor" href="#flowdiagram" aria-hidden="true"><span class="octicon octicon-link"></span></a>Flow diagram</h1>
<p>A general overview of various analysis steps involved in PARE.<br><br>

<img src="images/flowdiagram.png" border=0 width="55%" height="55%"></p>

<h1>
<a id="input" class="anchor" href="#input" aria-hidden="true"><span class="octicon octicon-link"></span></a>Input</h1>

<p>As input, the method requires two <a href="https://genome.ucsc.edu/goldenPath/help/bam.html" target="_blank">BAM</a> files corresponding to two replicates of H3K4me1 (enhancer prediction) or H3K4me3 (promoter prediction) ChIP-seq experiment.</p>

<p>PARE can be executed simply by issuing following command</p>

<pre><code>pare -i &lt;BAM file (replicate 1)&gt; -j &lt;BAM file (replicate 2)&gt; [OPTIONS]
</code></pre>

<h1>
<a id="output" class="anchor" href="#output" aria-hidden="true"><span class="octicon octicon-link"></span></a>Output</h1>

<p>The results from the PARESuite are presented in two text files:</p>

<pre><code>a) RESULTS.TXT: main result file in BED format. Each line represents the genomic coordinate of 
   an active enhancer or promoter.
</code></pre>

<p>For easy access, the html version of this file (RESULTS.HTML) is also available within the output directory</p>

<pre><code>b) RESULTS.UCSC: file to view the enhancer and promoter regions in UCSC browser.
</code></pre>

h1>
<a id="param" class="anchor" href="#param" aria-hidden="true"><span class="octicon octicon-link"></span></a>PARAMETERS</h1>
<pre><code>Program: pare (Predict Active Regulatory Elements using histone marks)
Author: BRIC, University of Copenhagen, Denmark
Version: 0.06
Contact: pundhir@binf.ku.dk
Usage: pare -i &lt;file(s)&gt; [OPTIONS]
Options:
    -i &lt;file&gt;   [mapped reads in BAM format (replicate 1)]
                [if multiple, please separate them by a comma]
[OPTIONS]
    -k &lt;file&gt;   [optimal histone peaks region in BED format]
                [-k will get precendence over -c parameter]
    -c &lt;file&gt;   [mapped reads in BAM format (control)]
    -o &lt;dir&gt;    [output directory to store results (default: results)
    -m &lt;string&gt; [genome (default: mm9)]
    -p          [run in parallel]
    -n &lt;int&gt;    [minimum length of nucleosome free region (default: 20)]
    -v &lt;int&gt;    [maximum length of nucleosome free region (default: 1000)]
    -t &lt;float&gt;  [FDR at which to consider a NFR as significant (default: 0.05)]
    -h          [help]
</code></pre>

<h1>
<a id="example" class="anchor" href="#example" aria-hidden="true"><span class="octicon octicon-link"></span></a>Example</h1>

<p>An usage example of PARESuite is shown below. As input, the method requires mapped reads in BAM format. An example dataset and expected results are available at <a href="http://servers.binf.ku.dk/pare/download/test_run/">http://servers.binf.ku.dk/pare/download/test_run/</a></p>

<pre><code>pare -i data/h3k4me1_helas3_Rep1.bam,data/h3k4me1_helas3_Rep2.bam -k \
data/optimal.h3k4me1_helas3_Rep0_Vs_control_helas3_Rep0.regionPeak.gz -o results \
-m hg19 -p &amp;&gt; pare.log
</code></pre>

<p>UCSC browser view of erythorid lineage (MEP) specific enhancers (in red color) identified through H3K4me1 PVP-pattern. Also shown is upregulation in the expression of gene, Rnf145 proximal to these enhancers in erythroid lineage (HSC: Hematopoietic Stem Cells; GMP: Granulocyte-Macrophage Progenitor; CLP: Common Lymphoid Progenitor; MEP: Megakaryocyte-Erythroid Progenitor).<br><br>

<img src="images/model_example.png" border=0 width="80%" height="80%"></p>

<h1>
<a id="more-info" class="anchor" href="#more-info" aria-hidden="true"><span class="octicon octicon-link"></span></a>More info</h1>

<p>For information on how to install PARE and its dependencies, please refer to <a href="https://github.com/spundhir/pare">https://github.com/spundhir/pare</a></p>

<h1>
<a id="license" class="anchor" href="#license" aria-hidden="true"><span class="octicon octicon-link"></span></a>License</h1>

<pre><code>PARE: a computational method to Predict Active Regulatory Elements using histone marks
Copyright (C) 2015  Sachin Pundhir (pundhir@binf.ku.dk)

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see &lt;http://www.gnu.org/licenses/&gt;.
</code></pre>

      <footer class="site-footer">
        <span class="site-footer-owner"><a href="https://github.com/spundhir/PARE">PARE</a> is maintained by <a href="https://github.com/spundhir">Sachin Pundhir</a>.</span>

        <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the <a href="https://github.com/jasonlong/cayman-theme">Cayman theme</a> by <a href="https://twitter.com/jasonlong">Jason Long</a>.</span>
      </footer>

    </section>

  
  </body>
</html>
